README.md
# glassmorphism_ui 

Create glassmorphic design using GlassContainer having frosted glass effect.

----------
## Getting Started

Create beautiful glassmorphic designs using this package,GlassContainer Widget provided by this package is completely customizable,change Container's opacity,blur,border,radius etc.


![alt text](./lib/glass.jpg)

### GlassContainer

```dart
GlassContainer(
  height: 200,
  width: 200,
  blur: 4,
  color: Colors.white.withOpacity(0.1),
  gradient: LinearGradient(
    begin: Alignment.topLeft,
    end: Alignment.bottomRight,
    colors: [
      Colors.white.withOpacity(0.2),
      Colors.blue.withOpacity(0.3),
    ],
  ),
  //--code to remove border
  border: Border.fromBorderSide(BorderSide.none),
  shadowStrength: 5,
  shape: BoxShape.circle,
  borderRadius: BorderRadius.circular(16),
  shadowColor: Colors.white.withOpacity(0.24),
),
```

### GlassImage

Image with frosted glass overlay

```dart
GlassImage(
  height: 200,
  width: 200,
  blur: 4,
  image: Image.network(
    "link",
    fit: BoxFit.cover,
  ),
  overlayColor: Colors.white.withOpacity(0.1),
  gradient: LinearGradient(
    begin: Alignment.topLeft,
    end: Alignment.bottomRight,
    colors: [
      Colors.white.withOpacity(0.2),
      Colors.blue.withOpacity(0.3),
    ],
  ),
  border: Border.fromBorderSide(BorderSide.none),
  shadowStrength: 5,
  borderRadius: BorderRadius.circular(16),
  shadowColor: Colors.white.withOpacity(0.24),
),
```

If you just want to blur image use `BlurredImage` widget

```dart
ClipRect(
  child: BlurredImage(
  image: Image.network(
      "link",
      fit: BoxFit.cover,
    ),
  ),
),
```

Or use extension method

```dart
Image.network(
  "https://cdn.pixabay.com/photo/2015/04/23/22/00/tree-736885__480.jpg",
  fit: BoxFit.cover,
).blur(blur: 4),

// Clip blur area
Image.network(
  "https://cdn.pixabay.com/photo/2015/04/23/22/00/tree-736885__480.jpg",
  fit: BoxFit.cover,
).blurClipped(blur: 4),
```

### Blur 

Widget that blur only it's child.

```dart
Blur(blur: 4, child: Text("Glassmorphism")),

//extension method
Text("Glassmorphism").blur(blur: 4),
```


CHANGELOG.md
## [0.3.0] 

* new widgets
* more properties added to GlassContainer

## [0.2.0] 

* null-safety migration

## [0.1.0] 

* initial release

lib/src/blur.dart
import 'dart:ui';

import 'package:flutter/material.dart';

/// Widget that blur it's child
class Blur extends StatelessWidget {
  /// Widget that blur it's child
  const Blur({super.key, required this.blur, required this.child});

  ///Blur intensity
  final double blur;

  ///[Widget] child to be blurred.
  final Widget child;

  @override
  Widget build(BuildContext context) {
    if (blur < 1) return child;

    return ImageFiltered(
      imageFilter: ImageFilter.blur(sigmaX: blur, sigmaY: blur),
      child: child,
    );
  }
}

///Widget Blur Extensions
extension BlurWidgetExt on Widget {
  ///Blur the widget
  Widget blur({required double blur}) {
    return Blur(blur: blur, child: this);
  }

  ///[Blur] widget with [ClipRect] parent
  Widget blurClipped({required double blur}) {
    return ClipRect(child: Blur(blur: blur, child: this));
  }
}


lib/glassmorphism.dart
library glassmorphism;

import 'dart:ui';
import 'package:flutter/material.dart';

///[GlassContainer] Glassmorphism Ui
///Container with frosted glass effect
///Completely Customizable ,change blur,border,radius,
///opacity to achieve desired effect
///
/// Note:
///It Inherit properties of [Container] so expect layout effect as container,
///while tinkering with height and width

class GlassContainer extends StatelessWidget {
  ///[opacity] is used to control the glass frosted effect
  ///[opacity] should be between 0 and 1
  ///1 means full opaque
  ///0 means full transparent
  ///default value is 0.1
  final double opacity;

  ///[Widget] [child]
  final Widget child;

  ///blur intensity
  ///default value is 5
  final double blur;

  /// shadow strength
  /// default value is 4
  /// to remove shadow provide value 0
  /// shadowStrength : 0
  final double shadowStrength;

  ///Border Radius
  ///example:
  ///borderRadius: BorderRadius.circular(10),
  ///default value is same as example
  ///to remove border use [BorderRadius.zero],
  final BorderRadiusGeometry borderRadius;

  ///[GlassContainer] Height in [double]
  ///no default height
  final double height;

  ///[GlassContainer] Width in [double]
  ///no default width
  final double width;

  ///[border] example
  ///Border.all(
  ///   color: Colors.white.withOpacity(0.3),
  ///   width: 0.3,
  ///   style: BorderStyle.solid,
  ///),
  ///default is same as above example
  final BoxBorder border;

  const GlassContainer(
      {Key key,
      this.opacity = 0.05,
      this.child,
      this.blur = 5,
      this.border,
      this.height,
      this.width,
      this.borderRadius,
      this.shadowStrength = 4})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: _PaintShadow(shadowStrength: shadowStrength),
      child: Container(
        height: height,
        foregroundDecoration: BoxDecoration(
          borderRadius: borderRadius ?? BorderRadius.circular(10),
          border: border ??
              Border.all(
                  color: Colors.white.withOpacity(0.3),
                  width: 0.3,
                  style: BorderStyle.solid),
        ),
        width: width,
        child: ClipRRect(
          borderRadius: borderRadius ?? BorderRadius.circular(10),
          child: BackdropFilter(
            filter: ImageFilter.blur(
              sigmaX: blur,
              sigmaY: blur,
            ),
            child: Container(
              decoration: BoxDecoration(
                borderRadius: borderRadius ?? BorderRadius.circular(10),
                color: Colors.grey[100].withOpacity(opacity),
              ),
              child: child,
            ),
          ),
        ),
      ),
    );
  }
}

class _PaintShadow extends CustomPainter {
  _PaintShadow({
    this.shadowStrength = 1,
  });

  final double shadowStrength;

  static double convertRadiusToSigma(double radius) {
    return radius * 0.57735 + 0.5;
  }

  Paint customPainter(
      {double blurStrength, Color color, double opacity, double strokeWidth}) {
    return Paint()
      ..style = PaintingStyle.stroke
      ..color = color.withOpacity(0.24)
      ..strokeWidth = strokeWidth
      ..maskFilter = MaskFilter.blur(
        BlurStyle.normal,
        convertRadiusToSigma(
          blurStrength,
        ),
      );
  }

  @override
  void paint(Canvas canvas, Size size) {
    if (shadowStrength == 0) return;
    RRect rect = RRect.fromRectAndRadius(
        Rect.fromPoints(
          Offset(-shadowStrength / 2, -shadowStrength / 2),
          Offset(
            size.width + shadowStrength / 2,
            size.height + shadowStrength / 2,
          ),
        ),
        Radius.circular(10));
    canvas.drawRRect(
      rect,
      customPainter(
        color: Color(0xff333333),
        blurStrength: 20,
        strokeWidth: shadowStrength,
        opacity: 0.24,
      ),
    );
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => false;
}


lib/src/glass_image.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:glassmorphism_ui/src/blur.dart';
import 'package:glassmorphism_ui/src/glass_container.dart';

/// [GlassImage]
///
/// Image with frosted glass overlay
class GlassImage extends StatelessWidget {
  ///```
  /// opacity is used to control the glass frosted effect
  ///
  /// Value should be in between 0 and 1
  ///
  /// No effect if color is set
  ///
  /// default value : 0.1
  ///```
  final double opacity;

  ///```
  ///blur intensity
  ///default value : 5
  ///```
  final double blur;

  ///```
  /// shadow strength
  ///
  /// default value : 4
  /// ```
  final double shadowStrength;

  ///```
  /// shadow color
  ///
  /// default value : 4
  /// ```
  final Color shadowColor;

  ///```
  ///borderRadius [BorderRadiusGeometry]
  ///
  ///example:
  ///BorderRadius.circular(10),
  ///
  /// default value : BorderRadius.circular(10),
  ///
  ///```
  final BorderRadius? borderRadius;

  ///[GlassContainer] Height
  final double? height;

  ///[GlassContainer] Width
  final double? width;

  ///[overlayColor]
  ///
  ///default value: Colors.grey.shade100.withOpacity(opacity)
  final Color? overlayColor;

  ///```
  ///border [BoxBorder]
  ///
  ///example:
  ///Border.all(
  ///   color: Colors.white.withOpacity(0.3),
  ///   width: 0.3,
  ///   style: BorderStyle.solid,
  ///),
  ///
  ///default is same as example
  ///```
  final BoxBorder? border;

  /// [Gradient] gradient
  ///
  /// [overlayColor] will be ignored if gradient is set.
  final Gradient? gradient;

  ///[Image]
  final Image image;

  /// [GlassImage]
  ///
  /// Image with frosted glass overlay
  const GlassImage({
    super.key,
    this.opacity = 0.05,
    this.blur = 5,
    this.border,
    this.height,
    this.width,
    this.borderRadius,
    this.shadowStrength = 2,
    this.shadowColor = const Color(0x3D333333),
    this.overlayColor,
    this.gradient,
    required this.image,
  });

  @override
  Widget build(BuildContext context) {
    final _borderRadius = borderRadius ?? BorderRadius.circular(10);

    return Stack(
      children: [
        SizedBox(
          height: height,
          width: width,
          child: ClipRRect(
            borderRadius: _borderRadius,
            child: ImageFiltered(
              imageFilter: ImageFilter.blur(sigmaX: blur, sigmaY: blur),
              child: image,
            ),
          ),
        ),
        GlassContainer(
          blur: 0,
          width: width,
          border: border,
          borderRadius: _borderRadius,
          color: overlayColor,
          gradient: gradient,
          opacity: opacity,
          shadowColor: shadowColor,
          shadowStrength: shadowStrength,
          height: height,
        ),
      ],
    );
  }
}

/// Blur Image
///
/// To blur any widget use [Blur] widget
///
/// Wrap it with [ClipRect] widget if it blur outside the widget or use [ImageExtension]'s blurClipped extension method
///
/// ```dart
/// ClipRect(
///  child: BlurredImage(
///   image: Image.network(
///      "link",
///      fit: BoxFit.cover,
///    ),
///   ),
/// ),
/// ```
class BlurredImage extends Blur {
  /// Blur Image
  const BlurredImage({super.key, super.blur = 4, required Image image})
      : super(child: image);
}

///[ImageExtension]
extension ImageExtension on Image {
  /// Create [BlurredImage]
  Widget blur({double blur = 3}) {
    return BlurredImage(image: this, blur: blur);
  }

  /// Create [BlurredImage] with [ClipRect] parent
  Widget blurClipped({double blur = 3}) {
    return ClipRect(child: BlurredImage(image: this, blur: blur));
  }
}


lib/src/glass_container.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:glassmorphism_ui/src/hollo_shadow_painter.dart';

///[GlassContainer] Container with frosted glass effect
class GlassContainer extends StatelessWidget {
  ///```
  /// opacity is used to control the glass frosted effect
  ///
  /// Value should be in between 0 and 1
  ///
  /// No effect if color is set
  ///
  /// default value : 0.1
  ///```
  final double opacity;

  ///[Widget] [child]
  final Widget? child;

  ///```
  ///blur intensity
  ///default value : 5
  ///```
  final double blur;

  ///```
  /// shadow strength
  ///
  /// default value : 4
  /// ```
  final double shadowStrength;

  ///```
  /// shadow color
  ///
  /// default value : 4
  /// ```
  final Color shadowColor;

  ///```
  ///borderRadius [BorderRadiusGeometry]
  ///
  ///example:
  ///BorderRadius.circular(10),
  ///
  /// default value : BorderRadius.circular(10),
  ///
  ///```
  final BorderRadius? borderRadius;

  ///[GlassContainer] Height
  final double? height;

  ///[GlassContainer] Width
  final double? width;

  ///[GlassContainer] color
  ///
  ///default value: Colors.grey.shade100.withOpacity(opacity)
  final Color? color;

  ///```
  ///border [BoxBorder]
  ///
  ///example:
  ///Border.all(
  ///   color: Colors.white.withOpacity(0.3),
  ///   width: 0.3,
  ///   style: BorderStyle.solid,
  ///),
  ///
  ///default is same as example
  ///```
  final BoxBorder? border;

  /// Container's shape
  final BoxShape shape;

  /// [Gradient] gradient
  ///
  /// [color] will be ignored if gradient is set.
  final Gradient? gradient;

  ///[GlassContainer] Container with frosted glass effect
  ///
  ///Note:
  ///
  ///It Inherit properties of [Container] so expect layout effect as container,
  ///while tinkering with height and width
  const GlassContainer({
    super.key,
    this.opacity = 0.05,
    this.child,
    this.blur = 2,
    this.border,
    this.height,
    this.width,
    this.borderRadius,
    this.shadowStrength = 4,
    this.shadowColor = const Color(0x3D333333),
    this.color,
    this.gradient,
    this.shape = BoxShape.rectangle,
  });

  @override
  Widget build(BuildContext context) {
    final _color = color ?? Colors.grey.shade100.withOpacity(opacity);
    final _borderRadius = borderRadius ?? BorderRadius.circular(10);
    final _border = border ??
        Border.all(
          color: Colors.white.withOpacity(0.3),
          width: 0.3,
        );

    Widget _child = DecoratedBox(
      decoration: BoxDecoration(
        shape: shape,
        color: _color,
        borderRadius: shape == BoxShape.circle ? null : _borderRadius,
        gradient: gradient,
      ),
      child: child,
    );

    if (blur > 0) {
      _child = ClipRRect(
        borderRadius: _borderRadius,
        child: BackdropFilter(
          filter: ImageFilter.blur(sigmaX: blur, sigmaY: blur),
          child: _child,
        ),
      );
    }

    return CustomPaint(
      painter: HollowShadowPainter(
        shape: shape,
        shadowStrength: shadowStrength,
        shadowColor: shadowColor,
        borderRadius: _borderRadius,
      ),
      child: Container(
        width: width,
        height: height,
        foregroundDecoration: BoxDecoration(
          borderRadius: _borderRadius,
          border: _border,
        ),
        child: _child,
      ),
    );
  }
}


lib/src/hollo_shadow_painter.dart
import 'package:flutter/material.dart';

///Render Shadow around the container
class HollowShadowPainter extends CustomPainter {
  HollowShadowPainter({
    required this.shape,
    required this.borderRadius,
    required this.shadowColor,
    this.shadowStrength = 1,
  });

  final double shadowStrength;
  final Color shadowColor;
  final BoxShape shape;
  final BorderRadius borderRadius;

  static double _convertRadiusToSigma(double radius) {
    return radius * 0.57735 + 0.5;
  }

  Paint customPainter({
    required double blurStrength,
    required Color color,
    required double strokeWidth,
  }) {
    return Paint()
      ..style = PaintingStyle.stroke
      ..color = color
      ..strokeWidth = strokeWidth
      ..maskFilter = MaskFilter.blur(
        BlurStyle.normal,
        _convertRadiusToSigma(blurStrength),
      );
  }

  @override
  void paint(Canvas canvas, Size size) {
    if (shadowStrength == 0) return;

    if (shape == BoxShape.circle) {
      return canvas.drawCircle(
        Offset(size.width / 2, size.height / 2),
        size.width / 2 + shadowStrength / 2,
        customPainter(
          color: shadowColor,
          blurStrength: 20,
          strokeWidth: shadowStrength,
        ),
      );
    }

    final RRect rrect = RRect.fromRectAndCorners(
      Rect.fromPoints(
        Offset(-shadowStrength / 2, -shadowStrength / 2),
        Offset(
          size.width + shadowStrength / 2,
          size.height + shadowStrength / 2,
        ),
      ),
      bottomLeft: borderRadius.bottomLeft,
      bottomRight: borderRadius.bottomRight,
      topLeft: borderRadius.topLeft,
      topRight: borderRadius.topRight,
    );

    canvas.drawRRect(
      rrect,
      customPainter(
        color: shadowColor,
        blurStrength: 20,
        strokeWidth: shadowStrength,
      ),
    );
  }

  @override
  bool shouldRepaint(HollowShadowPainter oldDelegate) => oldDelegate != this;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is HollowShadowPainter &&
        other.shadowStrength == shadowStrength &&
        other.shadowColor == shadowColor &&
        other.shape == shape &&
        other.borderRadius == borderRadius;
  }

  @override
  int get hashCode {
    return shadowStrength.hashCode ^
        shadowColor.hashCode ^
        shape.hashCode ^
        borderRadius.hashCode;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:glassmorphism_ui/glassmorphism_ui.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Glassmorphism ',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        decoration: BoxDecoration(
          gradient: LinearGradient(
            begin: Alignment.topLeft,
            end: Alignment.bottomRight,
            colors: [
              Colors.blue,
              Colors.pink,
            ],
          ),
          // color: Colors.red,
        ),
        child: Center(
          child: Stack(
            children: [
              Container(
                child: Center(
                  child: Text(
                    "Glassmorphism",
                    style: TextStyle(
                      color: Colors.white,
                      fontSize: 40,
                    ),
                  ),
                ),
              ),
              GlassImage(
                height: 200,
                width: 200,
                blur: 4,
                image: Image.network(
                  "link",
                  fit: BoxFit.cover,
                ),
                overlayColor: Colors.white.withOpacity(0.1),
                gradient: LinearGradient(
                  begin: Alignment.topLeft,
                  end: Alignment.bottomRight,
                  colors: [
                    Colors.white.withOpacity(0.2),
                    Colors.blue.withOpacity(0.3),
                  ],
                ),
                border: Border.fromBorderSide(BorderSide.none),
                shadowStrength: 5,
                borderRadius: BorderRadius.circular(16),
                shadowColor: Colors.white.withOpacity(0.24),
              ),
              Image.network(
                "https://cdn.pixabay.com/photo/2015/04/23/22/00/tree-736885__480.jpg",
                fit: BoxFit.cover,
              ).blur(blur: 4),
              Blur(blur: 4, child: Text("Glassmorphism")),
              Image.network(
                "https://cdn.pixabay.com/photo/2015/04/23/22/00/tree-736885__480.jpg",
                fit: BoxFit.cover,
              ).blurClipped(blur: 4),
              Center(
                child: ClipRect(
                  child: BlurredImage(
                    image: Image.network(
                      "https://cdn.pixabay.com/photo/2015/04/23/22/00/tree-736885__480.jpg",
                      fit: BoxFit.cover,
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
